QccPack-0.61-1/src/libQccPackWAV/utils/man1/imgrdwt.1
.TH IMGRDWT 1 "QCCPACK" ""
.SH NAME
imgrdwt, imgirdwt \- RDWT an image, and corresponding inverse
transform
.SH SYNOPSIS
.B imgrdwt
.RB "[\|" \-w
.IR  wavelet "\|]"
.RB "[\|" \-b
.IR  boundary "\|]"
.RB "[\|" \-nl
.IR  levels "\|]"
.I imgfile
.I icpfiles...
.LP
.B imgirdwt
.RB "[\|" \-w
.IR  wavelet "\|]"
.RB "[\|" \-b
.IR  boundary "\|]"
.I icpfiles...
.I imgfile
.SH OPTIONS
.TP
.BI \-w " wavelet"
String. 
Name of the wavelet file
.RB ( FBK
or
.B LFT
format); default = "CohenDaubechiesFeauveau.9-7.lft"
.TP
.BI \-b " boundary"
String. Type of boundary extension; default = "symmetric".
.TP 
.BI \-nl " levels"
Integer. Number of levels of dyadic decomposition; default value is determined
from number of output
.IR icpfiles
.SH DESCRIPTION
.LP
.B imgrdwt
implements a 2D separable redundant discrete wavelet transform (RDWT)
of grayscale image
.IR imgfile .
This wavelet decomposition is accomplished by applying a 1D
RDWT along every row and then along every column.
The output of this process is
a low-resolution approximation (the baseband image)
to the original image, as well as highpass subbands
in the horizontal, vertical, and diagonal directions.
Each of these subbands has the same size as the original image.
If
.I levels
is greater than 1, the baseband subband is recursively decomposed using this
same process.
.LP
.I wavelet
is the name of a file containing either the
filter bank 
.RB ( FBK 
format) or lifting scheme
.RB ( LFT
format) to use in performing the DWT.
The name of the filter-bank/lifting-scheme file is given by
.IR wavelet ;
this file must be found in the path given by the
environment variable
.BR QCCPACK_WAVELET_PATH .
The appropriate transform implementation
(filter bank, lifting) is used based upon the type
.RB ( FBK , " LFT" )
of the found file.
If the user does not specify a wavelet,
the default wavelet is a lifting implementation of the
Cohen-Feauveau-Daubechies 9/7 spline with less dissimilar lengths, a
biorthogonal wavelet popular wavelet for image-processing
applications (see Cohen 
.IR "et al" .,
Antonini 
.IR "et al" .).
.LP
.I boundary
gives the method for handling the image boundaries.  Possible values
for boundary are described in the page for
.BR QccWAVWavelet (3).
Not all boundary methods are permitted for
all wavelets; check the comments at the start of
.B FBK
or
.B LFT
files for possible values for particular wavelets.
The default boundary-extension method is "symmetric".
.LP
.IR levels
gives the number of transform decomposition levels.
Since at every scale, or level, of RDWT decomposition, four subbands are
produced, the total number of subbands produced in a transform with
.IR levels
levels of decomposition is
.RI "3 * " levels " + 1."
.B imgdwt
saves the resulting subbands to the 
.BR ICP -format
files
.IR icpfiles .
If
.IR levels
is specified,
.B imgrdwt
ensures that the number of output
.IR icpfiles
is less than or equal
.RI "3 * " levels " + 1."
On the other hand, if
.IR levels
is not specified,
.B imgrdwt
looks at how many
.I icpfiles
are given on the command line to determine the number of
levels 
of transform to take.
In this case,
.B imgrdwt
checks to see that the number of
.I icpfiles
correctly specifies a number of
.IR levels .
For example, if one gives 10
.I icpfiles
on the command line, then
.B imgrdwt
performs a 3-level RDWT,
since
.IR "levels" " =  (10 - 1) / 3 = 3."
On the other hand, if ones gives, say, 9 or 11
.I icpfiles
on the command line,
.B imgrdwt
exits in error, since neither 8 nor 10 are divisble by 3 and therefore
do not specify a possible number of transform scales.
.LP
.B imgirdwt
implements the corresponding inverse RDWT.
The number of levels of inverse RDWT to be performed is extracted
from the the number of input
.I icpfiles
specified on the command line
in a manner identical to that described above for
.BR imgrdwt .
.LP
.B imgrdwt
and
.B imgrdwt
are implemented via calls
to
.BR QccWAVWaveletRedundantDWT2D (3)
and
.BR QccWAVWaveletRedundantDWT2D (3),
respectively.
See the corresponding documentation for those functions
for details on the RDWT and its relation to the usual
critically sampled DWT as implemented by
.BR imgdwt (1).
.SH "SEE ALSO"
.BR QccWAVWaveletRedundantDWT2D (3),
.BR QccWAVWaveletInverseRedundantDWT2D (3),
.BR imgdwt (1),
.BR imgidwt (1),
.BR QccPackWAV (3),
.BR QccPackIMG (3),
.BR QccPack (3)

A. Cohen, I. Daubechies, J. C. Feauveau, "Biorthogonal Bases of
Compactly Supported Wavelets," 
.IR "Communications on Pure and Applied Mathematics" ,
vol. 45, no. 5, pp. 485-560, May 1992.

M. Antonini, M. Barlaud, P. Mathieu, I. Daubechies, "Image Coding Using
Wavelet Transform," 
.IR "IEEE Transactions on Image Processing" ,
vol. 1, no. 2, pp. 205-220, April 1992.

.SH AUTHOR
Copyright (C) 1997-2016  James E. Fowler
.\"  The programs herein are free software; you can redistribute them and/or
.\"  modify them under the terms of the GNU General Public License
.\"  as published by the Free Software Foundation; either version 2
.\"  of the License, or (at your option) any later version.
.\"  
.\"  These programs are distributed in the hope that they will be useful,
.\"  but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"  GNU General Public License for more details.
.\"  
.\"  You should have received a copy of the GNU General Public License
.\"  along with these programs; if not, write to the Free Software
.\"  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.


QccPack-0.61-1/CHANGELOG
==============================================================================

                                  Q C C P A C K

                            Version 0.61, 01-apr-2016

                    Copyright (C) 1997-2016  James E. Fowler

------------------------------------------------------------------------------


CHANGES:

Version 0.61:
  01-apr-2016:  BETA release version 0.61
  30-mar-2016:  Updated Cygwin installation instructions
  20-jan-2016:  Updated copyright year range for 2016
  20-jan-2016:  Fixed zero-subband-size bug in tce3d.c

Version 0.60:
  25-nov-2015:  BETA release version 0.60
  08-aug-2015:  Added QccFileReadLine()
  07-aug-2015:  Added QccFileGetRealPath()
  14-jul-2015:  Added QccFileWriteMagicNumberVersion()
  
Version 0.59:
  18-jun-2015:  BETA release version 0.59
  15-jun-2015:  Added QccFileRewind()
  14-jun-2015:  Updated copyright year range for 2015
  14-jun-2015:  Made QccENTExponentialGolombEncodeSymbol() and
                  QccENTExponentialGolombDecodeSymbol() visible
  09-jan-2014:  Updated copyright year range for 2014
  20-jul-2013:  Added Data/Matlab to QCCPACK_DATA_DIRS in Imakefile
  20-jul-2013:  Updated copyright year range for 2013
  21-dec-2011:  Updated copyright year range for 2011
  21-dec-2011:  Minor fixes for warning-free compilation with gcc-4.6
  23-jun-2009:  Added better support for uniform quantizers to sqencode,
                  sqdecode
  11-mar-2009:  Added logo.sed

Version 0.58:
  13-feb-2009:  BETA release version 0.58
  05-feb-2009:  Added klttcedecode3d.1, klttcedecode3d_lossless.1, 
                  klttceencode3d.1, klttceencode3d_lossless.1, 
                  tcedecode3d_lossless.1, tceencode3d_lossless.1
  05-feb-2009:  Added QccWAVtce3DLosslessEncode.3,
                  QccWAVtce3DLosslessDecodeHeader.3, 
                  QccWAVtce3DLosslessDecode.3, 
                  QccWAVklttce3DEncode.3, 
                  QccWAVklttce3DDecodeHeader.3, 
                  QccWAVklttce3DDecode.3, 
                  QccWAVklttce3DLosslessEncode.3, 
                  QccWAVklttce3DLosslessDecodeHeader.3, 
                  QccWAVklttce3DLosslessDecode.3
  05-feb-2009:  Added klttce3d.c, klttceencode3d.c, klttcedecode3d.c
  04-feb-2009:  Changed QccHYPrkltTrain(), QccHYPrkltTransform(),
                  QccHYPrkltInverseTransform() to use integer volumes
  04-feb-2009:  Added klttce3d_lossless.c, klttceencode3d_lossless.c,
                  klttcedecode3d_lossless.c
  04-feb-2009:  Fixed zero temporal_num_scales problem in 3D packet transforms
  02-feb-2009:  Added tce3d_lossless.c,
                  tceencode3d_lossless, tcedecode3d_lossless
  29-jan-2009:  QccHYPrkltTrain.3, QccHYPrkltFactorization.3,
                  QccHYPrkltTransform.3, QccHYPrkltInverseTransform.3
  29-jan-2009:  Added QccHYPrklt.3
  29-jan-2009:  Added QccHYPrkltInitialize(), QccHYPrkltAlloc(),
                  QccHYPrkltFree(), QccHYPrkltTrain(),
                  QccHYPrkltFactorization()
  28-jan-2009:  Added QccHYPrkltTransform(), QccHYPrkltInverseTransform()
  28-jan-2009:  Added QccMatrixIdentity(), QccMatrixRowExchange(),
                  QccMatrixColExchange(), QccMatrixIdentity.3,
                  QccMatrixRowExchange.3, QccMatrixColExchange.3
  16-jan-2009:  Updated copyright year range for 2009

Version 0.57:
  30-jun-2008:  BETA release version 0.57
  09-may-2008:  Added QccTimeTic(), QccTimeToc(), QccTime.3
  06-may-2008:  Fixed spatial_num_levels/temporal_num_levels bug in
                  initialization in bisk3d.c, wdr3d.c
  06-may-2008:  Fixed zero number of levels in
                  QccWAVWaveletPacketDWT3D(),
                  QccWAVWaveletInversePacketDWT3D()
  26-apr-2008:  Added sam.c, icbsam.c, QccHYPImageCubeMeanSAM(),
                QccVectorAngle()
  17-apr-2008:  Added QccWAVtce3DEncode.3, tceencode3d.1
  16-apr-2008:  Added tce3d.c, tceencode3d.c, tcedecode3d.c
  16-apr-2008:  Added -bpv to rawtoicb.c, icbtoraw.c, rawdist3d.c
  16-apr-2008:  Added -bpp to rawtoicp.c, icptoraw.c, rawdist2d.c
  16-apr-2008:  Added bpv, bpp to QccHYPRaw(), QccHYPRawRead2D(),
                  QccHYPRawWrite2D(), QccHYPRawDist2D(), QccHYPRawRead3D(),
                  QccHYPRawWrite3D(), QccHYPRawDist3D(),

Version 0.56:
  14-mar-2008:  BETA release version 0.56
  14-mar-2008:  Added QccIMGImageComponentLBT.3,
                  QccIMGImageComponentInverseLBT.3
  13-mar-2008:  Added QccFastDCT, QccFastDCTInitialize, QccFastDCTCreate,
                  QccFastDCTFree, QccFastDCTForwardTransform1D,
                  QccFastDCTInverseTransform1D, QccFastDCTForwardTransform2D,
                  QccFastDCTInverseTransform2D
  13-mar-2008:  Added fast_dct.c
  13-mar-2008:  Added dcttceencode.1, dcttcedecode.1
  13-mar-2008:  Added QccWAVdcttceEncode.3, QccWAVdcttceDecodeHeader.3,
                  QccWAVdcttceDecode.3
  13-mar-2008:  Renamed image_lapped_filter.c to image_lbt.c
  24-feb-2008:  Fixed memory leaks in dcttce.c, image_lapped_filter.c
  17-jan-2008:  Added Haar.fbk
  09-jan-2008:  Added dcttce.c, dcttceencode.c, dcttcedecode.c
  09-jan-2008:  Added image_lapped_filter.c
  09-jan-2008:  Added QccMatrixInverse(), QccMatrixInverse.3
  04-jan-2008:  Updated substr
  04-jan-2008:  Updated copyright year range for 2008
  11-dec-2007:  Fixed minor max_coefficients bug in QccWAVtceEncodeDWT()
  24-oct-2007:  Changed spawn to be able to handle commands with arguments
  09-oct-2007:  Added QCCMATHEPS
  01-oct-2007:  Added QccMatrixIntMultiply(), QccMatrixIntMultiply.3
  01-oct-2007:  Added QccMatrixMultiply(), QccMatrixMultiply.3
  28-sep-2007:  Added QccMatrixNullspace(), QccMatrixNullspace.3
  28-sep-2007:  Added QccMatrixOrthogonalize(), QccMatrixOrthogonalize.3
  28-sep-2007:  Modified QccMatrixSVD() to work with matrics of any size

Version 0.55:
  29-aug-2007:  BETA release version 0.55
  03-jul-2007:  Added limits.h include
  17-may-2007:  Fixed num_symbols bug in QccENTHuffmanDesign()

QccPack-0.61-1/src/libQccPackWAV/utils/imgrdwt.c
/*
 *
 * QccPack: Quantization, compression, and coding utilities
 * Copyright (C) 1997-2016  James E. Fowler
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

#include "imgrdwt.h"

#define USG_STRING "[-w %s:wavelet] [-b %s:boundary] [-nl %d:levels] %s:imgfile %*s:icpfiles..."

QccWAVWavelet Wavelet;
QccString WaveletFilename = QCCWAVWAVELET_DEFAULT_WAVELET;
QccString Boundary = "symmetric";

int NumFiles;
QccIMGImage InputImage;
QccIMGImageComponent *OutputImageComponents;
QccString *Filenames;

int NumLevels = -1;
int NumSubbands;

int ImageNumCols;
int ImageNumRows;


int main(int argc, char *argv[])
{
  int current_file;
  QccMatrix *matrices;

  QccInit(argc, argv);

  QccIMGImageInitialize(&InputImage);
  QccWAVWaveletInitialize(&Wavelet);

  if (QccParseParameters(argc, argv,
                         USG_STRING,
                         WaveletFilename,
                         Boundary,
                         &NumLevels,
                         InputImage.filename,
                         &NumFiles,
                         &Filenames))
    QccErrorExit();

  if (NumLevels < 0)
    {
      NumSubbands = NumFiles;
      NumLevels = QccWAVSubbandPyramidNumSubbandsToNumLevels(NumSubbands);
      if (!NumLevels)
        {
          QccErrorAddMessage("%s: Invalid number of output files",
                             argv[0]);
          QccErrorExit();
        }
    }
  else
    {
      NumSubbands = QccWAVSubbandPyramidNumLevelsToNumSubbands(NumLevels);
      if (NumFiles > NumSubbands)
        {
          QccErrorAddMessage("%s: Too many output files for a %d-level transform (use %d or fewer output files)",
                             argv[0],
                             NumLevels,
                             NumSubbands);
          QccErrorExit();
        }
    }

  if (QccWAVWaveletCreate(&Wavelet, WaveletFilename, Boundary))
    {
      QccErrorAddMessage("%s: Error calling QccWAVWaveletCreate()",
                         argv[0]);
      QccErrorExit();
    }
  
  if (QccIMGImageRead(&InputImage))
    {
      QccErrorAddMessage("%s: Error calling QccIMGImageRead()",
                         argv[0]);
      QccErrorExit();
    }
  if (QccIMGImageColor(&InputImage))
    {
      QccErrorAddMessage("%s: Color images not supported",
                         argv[0]);
      QccErrorExit();
    }

  ImageNumCols = InputImage.Y.num_cols;
  ImageNumRows = InputImage.Y.num_rows;

  if ((OutputImageComponents = 
       (QccIMGImageComponent *)
       malloc(NumSubbands * sizeof(QccIMGImageComponent))) == NULL)
    {
      QccErrorAddMessage("%s: Error allocating memory",
                         argv[0]);
      QccErrorExit();
    }
  
  if ((matrices =
       (QccMatrix *)malloc(NumSubbands * sizeof(QccMatrix))) == NULL)
    {
      QccErrorAddMessage("%s: Error allocating memory",
                         argv[0]);
      QccErrorExit();
    }

  for (current_file = 0; current_file < NumSubbands; current_file++)
    {
      QccIMGImageComponentInitialize(&(OutputImageComponents[current_file]));
      OutputImageComponents[current_file].num_rows = ImageNumRows;
      OutputImageComponents[current_file].num_cols = ImageNumCols;

      if (QccIMGImageComponentAlloc(&(OutputImageComponents[current_file])))
        {
          QccErrorAddMessage("%s: Error calling QccIMGImageComponentAlloc()",
                             argv[0]);
          QccErrorExit();
        }

      matrices[current_file] = OutputImageComponents[current_file].image;
    }
  
  for (current_file = 0; current_file < NumFiles; current_file++)
    QccStringCopy(OutputImageComponents[current_file].filename,
                  Filenames[current_file]);

  if (QccWAVWaveletRedundantDWT2D(InputImage.Y.image,
                                  matrices,
                                  ImageNumRows,
                                  ImageNumCols,
                                  NumLevels,
                                  &Wavelet))
    {
      QccErrorAddMessage("%s: Error calling QccWAVWaveletRedundantDWT2D()",
                         argv[0]);
      QccErrorExit();
    }
  
  for (current_file = 0; current_file < NumFiles; current_file++)
    {
      if (QccIMGImageComponentSetMaxMin(&(OutputImageComponents[current_file])))
        {
          QccErrorAddMessage("%s: Error calling QccIMGImageComponentSetMaxMin()",
                             argv[0]);
          QccErrorExit();
        }

      if (QccIMGImageComponentWrite(&(OutputImageComponents[current_file])))
        {
          QccErrorAddMessage("%s: Error calling QccIMGImageComponentWrite()",
                             argv[0]);
          QccErrorExit();
        }
    }

  for (current_file = 0; current_file < NumSubbands; current_file++)
    QccIMGImageComponentFree(&(OutputImageComponents[current_file]));
  QccFree(OutputImageComponents);
  QccFree(matrices);

  QccIMGImageFree(&InputImage);
  QccWAVWaveletFree(&Wavelet);

  QccExit;
}
